File: chaos/src/Geometry.scala
trait Geometry[T]
{
	val aspectRatio:Double
	val size:Vector[T]
	val frame:Frame[T]
}

object Geometry
{
	def apply(fr:Frame[Double]) : Geometry[Double] = {
		new Geometry[Double]{
			val frame = fr
			val size  = fr.max operate(fr.min,{ (l,r) => Math.abs(l-r)+0.000001 } )
			val aspectRatio = size.x / size.y
		}
	}
	
	def apply(dim:Vector[Int]) : Geometry[Int] = {
		new Geometry[Int]{
			val frame = Frame[Int]( Vector(0,0) , Vector(dim.x-1,dim.y-1) )
			val size  = dim
			val aspectRatio = size.x.asInstanceOf[Double] / size.y.asInstanceOf[Double]
		}
	}
	
	// geom     : picture geometry
	// dataGeom : target data geometry
	def transform(	imgGeom:Geometry [Int]    , 
					dataGeom:Geometry[Double] ):(Vector[Double])=>Vector[Double] = {
		//
		val ratio=	if (dataGeom.aspectRatio < imgGeom.aspectRatio)
						imgGeom.size.y.asInstanceOf[Double] / dataGeom.size.y
					else
						imgGeom.size.x.asInstanceOf[Double] / dataGeom.size.x
		
		val origin = dataGeom.frame.min map { _ *ratio }
		val offset:Vector[Double]=if( dataGeom.aspectRatio < imgGeom.aspectRatio  ){
			//data     canvas
			// **      *****
			// ** ---> *****
			// **      *****
			(	- origin.x + (imgGeom.size.x - (dataGeom.size.x*ratio))*0.5 ,
				- origin.y )
		}
		else {
			//data     canvas
			// *****      **
			// ***** ---> **
			// *****      **
			(	- origin.x ,
				- origin.y + (imgGeom.size.y - (dataGeom.size.y*ratio))*0.5 )
		}
		// println("ratio = "+ratio+" , (offsetX,offsetY)="+(offsetX,offsetY))
		
		{ p =>
			val q = ( p map { _ * ratio } ) operate(offset,_ + _)
			(q.x , imgGeom.size.y - q.y) }
	}
}

File: lua/geometry.lua
require "vector"
require "frame"

Geometry = {}
Geometry_mt = {__index = Geometry}

function Geometry:new(param_frame)
	assert(param_frame)
	local o = {}
	o.frame = param_frame
	o.size  = o.frame:size()
	assert( o.size.x>0 )
	assert( o.size.y>0 )
	o.aspect_ratio = o.size.x / o.size.y
	return setmetatable(o,Geometry_mt)
end

Geometry_mt.__tostring = function(a)
	return string.format("(frame=%s,size=%s,aspect_ratio=%f)",
				tostring(a.frame) ,
				tostring(a.size)  ,
				a.aspect_ratio )
end

-- self ... canvas
function Geometry:canvas_transform(param_data_geom)
	assert(self)
	local img_geom = self
	--
	-- inflate data geometry
	local c = (param_data_geom.frame.min + param_data_geom.frame.max ) / 2
	local s = param_data_geom.size * 1.1
	local data_geom=Geometry:new(Frame:new( c - s/2 , c + s/2 ) )
	--
	local ratio = 0.0
	if(data_geom.aspect_ratio < img_geom.aspect_ratio)then
		ratio = img_geom.size.y / data_geom.size.y
	else
		ratio = img_geom.size.x / data_geom.size.x
	end
	assert(ratio>0.0)

	local origin = data_geom.frame.min * ratio
	local offset = -origin
	if ( data_geom.aspect_ratio < img_geom.aspect_ratio  ) then
		-- data     canvas
		--  **      *****
		--  ** ---> *****
		--  **      *****
		offset.x = offset.x + (img_geom.size.x - (data_geom.size.x*ratio))*0.5
	else
		-- data     canvas
		--	*****      **
		--	***** ---> **
		--	*****      **
		offset.y = offset.y + (img_geom.size.y - (data_geom.size.y*ratio))*0.5
	end

	return function(p)
			local q = p*ratio + offset
			q.y = img_geom.size.y - q.y
			return q
		end
	end



